Port qemu forkserver version to v1

nj00001 opened the pull request:
In addition to upgrading the forkserver version, some ijon support code has been added. However, the afl-fuzz part also needs to be modified, mainly to expand the bitmap, but it is a bit complicated.

nj00001 committed:
Port qemu forkserver version to v1

In addition to upgrading the forkserver version, some ijon support code has been added. However, the afl-fuzz part also needs to be modified, mainly to expand the bitmap, but it is a bit complicated.

vanhauser-thc commented on accel/tcg/cpu-exec.c:
```diff
@@ -281,6 +281,15 @@ static void restore_memory_snapshot(void) {
+
+static void qemu_ijon_init() {
+  use_ijon = !!getenv("AFL_QEMU_IJON");
+  if (use_ijon == 0) return;
```
that line is pointless :-)

nj00001 commented:
> You could just add an AFL_QEMU_MAP_SIZE=xxxx and set this size in the qemu code for the map. but beware, you may only allow x^2 numbers, so 2^7, 2^8, x^9 ...


On the afl-fuzz side, the map_size set by qemu mode seems to be fixed at 64k. If the map_size passed by qemu through forkserver is larger than 64k, an error will be reported.

https://github.com/AFLplusplus/AFLplusplus/blob/b5b5af1becbc31cdb03de508f841f7abc8edd1c0/src/afl-fuzz.c#L2570
<img width="734" height="261" alt="image" src="https://github.com/user-attachments/assets/17fba520-524a-4fce-9c09-b10f32a13d4f" />

<img width="716" height="604" alt="image" src="https://github.com/user-attachments/assets/9bc047a9-0975-4695-aef8-fbd58bf25d0e" />

nj00001 committed:
Add ijon related functions, which requires updates to the afl-fuzz repository

Updated forkserver code and ijon code. Next, we will add yaml parsing function to read user ijon settings from the configuration file.

Commit:
add qemu ijon support

nj00001 commented:
Obviously json parsing is easier

vanhauser-thc commented:
before you implement a lot, maybe lets discuss how you imagine this can work?

nj00001 commented:
The environment variable indicates the file location, which contains a JSON array. The ijon component uses the data in the array to insert a tcg helper call at the specified code location. This external call runs the ijon method specified in the array, reads the global variable addresses of the target program specified in the array, and passes the results to the ijon function. The array currently contains [
["0x1234","ijon_max","0x1212","4"],
["0x9999","ijon_min","0x2000","8"]
]
The first element is the code address (the program under test), the second is the ijon method, the third is the variable address to be read, and the fourth is the variable length.

vanhauser-thc commented:
I find json overkill for this - why not just the same format as you have it in the array? still parsing, but easier to write for humans? but up to you.
for the variable address part - it could also be present in a register, do you want to support this?

nj00001 commented:
Indeed, there is no need to use json, it should be easy to add reading of registers

nj00001 commented:
<img width="1544" height="745" alt="image" src="https://github.com/user-attachments/assets/d9ff6428-6149-4220-9a3f-589316f0238b" />
Tested a few times, seems to be working properly

Commit:
Simplified the ijon configuration file format and added support for register data

nj00001 commented:
I tested it locally, and it works fine using either memory addresses or register names. Adding AFL_DEBUG=1 should also allow you to see ijon log messages.

vanhauser-thc commented on accel/tcg/cpu-exec.c:
```diff
@@ -281,6 +289,527 @@ static void restore_memory_snapshot(void) {
+uint32_t g_var_len[0x1000];
+uint32_t ijon_type[0x1000];
+
+/* 读取整个文件到 malloc 的缓冲区（调用者负责 free） */
```
no chinese character please

vanhauser-thc replied:
all occurances please :-)

vanhauser-thc commented on accel/tcg/cpu-exec.c:
```diff
@@ -281,6 +289,527 @@ static void restore_memory_snapshot(void) {
+                (unsigned long)g_var_addr[i],
+                (unsigned)g_var_len[i]);
+    }
+
```
what do you want to do if ijon_hooker_cnt is 0? exit or set use_ijon to 0?

vanhauser-thc commented on accel/tcg/cpu-exec.c:
```diff
@@ -281,6 +289,527 @@ static void restore_memory_snapshot(void) {
+  for (char *p = s; *p; ++p) *p = (char)toupper((unsigned char)*p);
+
+  /* 常见寄存器别名（RAX/EAX -> R_EAX 等） */
+  if (strcmp(s, "RAX") == 0 || strcmp(s, "EAX") == 0) { free(buf); return R_EAX; }
```
a) your code is case sensitive, why not use stracsecmp instead?
b) I think this will break compiling qemu for any other platform than amd64 (because I doubt that R_ECX is defined e.g. for arm targets)

vanhauser-thc commented on accel/tcg/cpu-exec.c:
```diff
@@ -281,6 +289,527 @@ static void restore_memory_snapshot(void) {
+                if (parse_uint64_str(f3, &gaddr) != 0) ok = 0;
+            } else {
+                uint32_t reg_enum = ijon_reg_to_addr(f3);
+                if (reg_enum == -1) ok = 0;
```
you should warn if any of the conversions fail and ok is set to 0, especially for the register check

vanhauser-thc commented on accel/tcg/cpu-exec.c:
```diff
@@ -643,56 +1174,91 @@ void afl_forkserver(CPUState *cpu) {
+    if (lkm_snapshot) status |= FS_OPT_SNAPSHOT;
+    if (sharedmem_fuzzing) status |= FS_NEW_OPT_SHDMEM_FUZZ;
+
+    u32 __afl_map_size = MAP_SIZE;
```
you hardcode the map size to what is in config.h same as we had previously. I think it would be nice if this could be changed at runtime, e.g. with AFL_QEMU_MAP_SIZE=.... - but this would require a check that the value - 1 is 0x...ffffff without any holes.

nj00001 replied:
It looks like there are a few hard limits in the code, I'll leave that for later.

vanhauser-thc commented:
I left a few comments, otherwise this looks good.

One more thing: you still import headers for json but you removed the json format parsing.

Commit:
Optimized the parsing logic of the ijon configuration file. If an error occurs, the error reason will be printed and the system will exit. Fixed the problem of register address acquisition function on different architectures.

Commit:
Add ijon hook support for arm and aarch64

Commit:
Using macros to simplify the hook code should make it easy to add IJON support to other architectures in the future.
